src/core/project/qgsproject.cpp

```diff
@@ -1305,6 +1308,87 @@ void QgsProject::setAvoidIntersectionsMode( const Qgis::AvoidIntersectionsMode m
   emit avoidIntersectionsModeChanged();
 }
 
+static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFlags projectReadFlags, Qgis::ProjectFlags projectFlags )
+{
+  QgsMapLayer::ReadFlags layerFlags = QgsMapLayer::ReadFlags();
+  if ( projectReadFlags & Qgis::ProjectReadFlag::DontResolveLayers )
+    layerFlags |= QgsMapLayer::FlagDontResolveLayers;
+  // Propagate trust layer metadata flag
+  if ( ( projectFlags & Qgis::ProjectFlag::TrustStoredLayerStatistics ) || ( projectReadFlags & Qgis::ProjectReadFlag::TrustLayerMetadata ) )
+    layerFlags |= QgsMapLayer::FlagTrustLayerMetadata;
+  // Propagate open layers in read-only mode
+  if ( ( projectReadFlags & Qgis::ProjectReadFlag::ForceReadOnlyLayers ) )
+    layerFlags |= QgsMapLayer::FlagForceReadOnly;
+
+  return layerFlags;
+}
+
+void QgsProject::preloadProviders( const QVector<QDomNode> &asynchronusLayerNodes,
+                                   const QgsReadWriteContext &context,
+                                   QMap<QString, QgsDataProvider *> &loadedProviders,
+                                   QgsMapLayer::ReadFlags layerReadFlags,
+                                   int totalProviderCount )
+{
+
+  QVector<QgsRunnableProviderCreator *> runnables;
+  QThreadPool threadPool;
+
+  QgsSettings settings;
+  if ( settings.contains( QStringLiteral( "provider-preloading/max-thread-count" ) ) )
+  {
+    threadPool.setMaxThreadCount( settings.value( QStringLiteral( "provider-preloading/max-thread-count" ) ).toInt() );
+  }
+  else
+  {
+    threadPool.setMaxThreadCount( QThread::idealThreadCount() );
+    settings.setValue( QStringLiteral( "provider-preloading/max-thread-count" ), threadPool.maxThreadCount() );
+  }
+
+  int i = 0;
+  int validLayerCount = 0;
+  QEventLoop *loop = new QEventLoop( this );
```

Comment:
Can't we have the event loop just as an ordinary local variable - `QEventLoop loop;` ? that would be cleaner, and used in various places